cognite/client/data_classes/agents/chat.py

```diff
@@ -123,6 +366,148 @@ class MessageList(CogniteResourceList[Message]):
+    action_id: str
+    role: Literal["action"] = "action"
+
+    @classmethod
+    def _load(cls, data: dict[str, Any], cognite_client: CogniteClient | None = None) -> ActionResult:
+        """Dispatch to the correct concrete action result class based on `type`."""
+        action_type = data.get("type", "")
+        action_class = _ACTION_RESULT_CLS_BY_TYPE.get(action_type, UnknownActionResult)
+        return action_class._load_result(data, cognite_client)
+
+    @abstractmethod
+    def dump(self, camel_case: bool = True) -> dict[str, Any]:
+        """Dump the action result to a dictionary."""
+        ...
+
+    @classmethod
+    @abstractmethod
+    def _load_result(cls, data: dict[str, Any], cognite_client: CogniteClient | None = None) -> ActionResult:
+        """Create a concrete action result instance from raw data."""
+        ...
+
+
+@dataclass
+class ClientToolResult(ActionResult):
+    """Result of executing a client tool, for sending back to the agent.
+
+    Args:
+        action_id (str): The ID of the action being responded to.
+        content (str | MessageContent): The result of executing the action.
+        data (list[Any] | None): Optional structured data.
+    """
+
+    _type: ClassVar[str] = "clientTool"
+    content: MessageContent = field(init=False)
+    data: list[Any] | None = field(init=False, default=None)
+
+    def __init__(
+        self,
+        action_id: str,
+        content: str | MessageContent,
+        data: list[Any] | None = None,
+    ) -> None:
+        self.action_id = action_id
+        if isinstance(content, str):
+            self.content = TextContent(text=content)
+        else:
+            self.content = content
+        self.data = data
+        self.role = "action"
+
+    def dump(self, camel_case: bool = True) -> dict[str, Any]:
+        return {
+            "role": self.role,
+            "type": self._type,
+            "actionId" if camel_case else "action_id": self.action_id,
+            "content": self.content.dump(camel_case=camel_case),
+            "data": self.data or [],
+        }
+
+    @classmethod
+    def _load_result(cls, data: dict[str, Any], cognite_client: CogniteClient | None = None) -> ClientToolResult:
+        content = MessageContent._load(data["content"])
+        return cls(
+            action_id=data["actionId"],
+            content=content,
+            data=data.get("data"),
+        )
+
+
+@dataclass
+class ToolConfirmationResult(ActionResult):
+    """Result of a tool confirmation request.
+
+    Args:
+        action_id (str): The ID of the action being responded to.
+        status (Literal["ALLOW", "DENY"]): Whether to allow or deny the tool execution.
+    """
+
+    _type: ClassVar[str] = "toolConfirmation"
+    status: Literal["ALLOW", "DENY"] = field(init=False)
+
+    def __init__(self, action_id: str, status: Literal["ALLOW", "DENY"]) -> None:
+        self.action_id = action_id
+        self.status = status
+        self.role = "action"
```

Comment:
I have not seen this before. A trick to allow required properties on a dataclass derived from a class with optional properties.